### PR TITLE
Add guidance on logging to CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by DaveCTurner**
Logging is important, but today we don't have any written guidance about
how we do it in Elasticsearch so contributors struggle to make an
informed decision about how to do it right. This commit adds some
guidance on logging to the contribution guide.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -456,6 +456,172 @@ checks if contributed code does not have the appropriate license headers.
...
+Elasticsearch uses Log4J for logging. In most cases you should log via a
+`Logger` named after the class that is writing the log messages, which you can
+do by declaring a static field of the class as follows:
+
+    private static final Logger logger = LogManager.getLogger(MyClassName.class);
+
+There are a number of other acceptable ways to obtain a `Logger`. For example,
+if there will be many instances of a long-lived component whose logs should be
+distinguished then each instance can use a prefixed logger obtained from one of
+the methods in `org.elasticsearch.common.logging.Loggers`.
```
Comment (gwbrown) [inline CONTRIBUTING.md:476]: An example might be useful here, especially considering 1) this document is going to be consumed by folks who are just starting and not yet familiar with our patterns, and 2) we don't actually use prefixed logger that much compared to regular loggers.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -456,6 +456,172 @@ checks if contributed code does not have the appropriate license headers.
...
+
+You can also use a `Supplier<ParameterizedMessage>` to avoid constructing
+expensive messages that will usually be discarded:
+
+    logger.trace(() -> new ParameterizedMessage("Tolstoy says [{}]", warAndPeace.toString()));
+
+Logging is an important behaviour of the system and all but the most trivial
+logging should have corresponding tests. These tests typically use a
+`org.elasticsearch.test.MockLogAppender` to capture the logs and assert that
+they are as expected.
```
Comment (gwbrown) [inline CONTRIBUTING.md:503]: This might be true in the abstract, but I've only ever seen/written logging tests for particularly important log messages, or ones that are complex to trigger and I want to make sure are working right. I'm mildly concerned the phrasing here overemphasizes the degree we expect logging tests for most instances of logging.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -456,6 +456,172 @@ checks if contributed code does not have the appropriate license headers.
...
+If an exception occurs as a direct result of a request received from a client
+then you should include the exception in the response back to the client rather
+than recording it in the logs at `WARN` level. The person reading the logs is
+usually unable to address any problems caused by faulty client requests, and
+the person running the client is often forbidden from seeing the server logs.
+
+##### `ERROR` and `FATAL`
+
+These are the least verbose levels, but they are not meaningfully different
+from `WARN` within Elasticsearch and you should almost never use them.
```
Comment (gwbrown) [inline CONTRIBUTING.md:623]: I think it's worth adding a bit more context here - even if this is our desired usage, there are several hundred uses of `.error()`, but just a single non-test use of `.fatal()` currently in the Elasticsearch codebase.

Is `ERROR`-level logging something we should be avoiding in new code? I haven't heard of any such desire, but it's very possible I'm just out of the loop.

**Review by AthenaEryma — COMMENTED** [review]
Left a few comments, but overall looks good - I think this will be super helpful for new joiners and in keeping expectations and usage consistent across the team.

File: CONTRIBUTING.md (hunk 3 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md:623]: Perhaps I'm just using [Cunningham's law](https://meta.wikimedia.org/wiki/Cunningham%27s_Law) here :)

I've seen folks get very worried by `ERROR`-level logging emitted by components on the grounds that log4j documents this level as for events which "might still allow the application to continue running" implying that there's a possibility that Elasticsearch actually _won't_ continue running after emitting such a log. There's not zero errors that fit that description, but it's a vanishingly small subset and almost none of the usages of this level fit that interpretation today.

File: CONTRIBUTING.md (hunk 2 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md:503]: ++ I write quite a lot of such tests but I'm willing to believe it's not that common. I'll weaken the language here.

File: CONTRIBUTING.md (hunk 1 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md:476]: The lack of an example was deliberate here. Almost everyone will do the thing I already suggested. Maybe it'd be better to have even fewer words here and just link to the `Loggers` class. You can always navigate to the class and see how/where it's used if you're interested.

File: CONTRIBUTING.md (hunk 3 above)
Comment (gwbrown) [reply CONTRIBUTING.md:623]: Perhaps a discussion to promote back to the wider ES team - I'm certainly not opposed to reducing our usage of `ERROR`-level logs, and a clearer definition of when to use them. I do think there are some uses for it which aren't necessarily fatal but are more severe than a warning. 

For example, it's possible (mostly through misconfiguration nowadays) for ILM to fail to roll over an index, and if that goes "untreated", the index is likely to continue to grow until its shards are big enough to start causing serious problems with the cluster. It's not immediately fatal, but represents a problem that needs to be handled sooner rather than later and using different severity level to indicate that may be useful.

Basically, I'm advocating for EITHER saying `Generally use neither ERROR nor FATAL, and if you see either re-evaluate if it should really be at that severity and adjust it if necessary` OR `Use ERROR for cases fitting these criteria: [...] and use FATAL only for errors which actually kill the process`. But just saying "Don't use `ERROR` or `FATAL`" when we use `ERROR` a bunch in existing code is going to be confusing.

File: CONTRIBUTING.md (hunk 2 above)
Comment (gwbrown) [reply CONTRIBUTING.md:503]: I can very easily believe the logging you write falls into the "particularly important or complex to trigger" camp much more often than the logging I write :)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -456,6 +456,171 @@ checks if contributed code does not have the appropriate license headers.
...
+
+If you wish to use placeholders and an exception at the same time, construct a
+`ParameterizedMessage`:
+
+    logger.debug(new ParameterizedMessage("failed at offset [{}]", offset), exception);
+
+You can also use a `Supplier<ParameterizedMessage>` to avoid constructing
+expensive messages that will usually be discarded:
+
+    logger.trace(() -> new ParameterizedMessage("rarely seen output [{}]", expensiveMethod()));
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply CONTRIBUTING.md]: yep, done in f4fa7570d85.

File: CONTRIBUTING.md (hunk 3 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md]: Ok I've reworded it to be more permissive around `ERROR`, more highlighting the consequences of using it than saying not to use it ever.

Comment (DaveCTurner) [reply CONTRIBUTING.md]: We could also bring it back to the team for further discussion, I just think that as written now it should be acceptable.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -456,6 +456,188 @@ checks if contributed code does not have the appropriate license headers.
...
+isolation.
+
+Even though `TRACE` logs may be very verbose, you should still exercise some
+judgement when deciding when to use them. In many cases it will be easier to
+understand the behaviour of the system using tests or by analysing the code
+itself rather than by trawling through hundreds of trivial log messages.
+
+It may not be easy, or even possible, to obtain `TRACE` logs from a production
+system. Therefore they are not appropriate for information that you would
+expect to be useful in diagnosing problems in production.
```
Comment (nik9000) [inline CONTRIBUTING.md:538]: I'm kind of unsure why you'd commit logs for some reason *other* than to debug stuff in production. I guess to debug test failures? I feel like TRACE logs are useful to debug some issues in production. But super rarely. At least, that makes sense given all the other things you mentioned - them being difficult to access in production and very very chatty.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -456,6 +456,188 @@ checks if contributed code does not have the appropriate license headers.
...
+
+It's possible that the reader of `DEBUG` logs is also reading the code, but
+that is less likely than for `TRACE` logs. You should strive to avoid
+terminology that only makes sense when reading the code, and also aim for
+messages at this level to be self-contained rather than intending them to be
+read as a sequence.
+
+It's often useful to log exceptions and other deviations from the "happy path"
+at `DEBUG` level. Exceptions logged at `DEBUG` should generally include the
+complete stack trace.
```
Comment (nik9000) [inline CONTRIBUTING.md:564]: It might be useful to call out a few things that shouldn't be logged at DEBUG level.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -456,6 +456,188 @@ checks if contributed code does not have the appropriate license headers.
...
+Although the stack trace may not be useful to the user, it may contain
+information that is vital for a developer to fully understand the problem. In
+other cases it may be appropriate to log the exception message at `WARN` and
+only log the full exception if the user has enabled `DEBUG` logging:
+
+    if (logger.isDebugEnabled()) {
+        logger.warn("investigate me", exception);
+    } else {
+        logger.warn("investigate me: [{}]", exception.getMessage())
+    }
```
Comment (nik9000) [inline CONTRIBUTING.md:605]: Do we only do this when we think that the stack trace is not generally useful? Because the cause is super well understood, I guess. It's kind of bothered me that you have to read the code to know that setting the thing to debug will cause the whole stack trace to be logged.

Anyway! I think that this pattern is a bit sneaky and I'd love to see examples of where we use it.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -456,6 +456,188 @@ checks if contributed code does not have the appropriate license headers.
...
+
+In a situation where occasional transient failures are expected and handled,
+but a persistent failure requires the user's attemption, consider implementing
+a mechanism to detect that a failure is unacceptably persistent and emit a
+corresponding `WARN` log. For example, it may be helpful to log every tenth
+consecutive failure at `WARN` level, or log at `WARN` if an operation has not
+completed within a certain time limit. This is much more user-friendly than
+failing persistently and silently by default and requiring the user to enable
+`DEBUG` logging to investigate the problem.
```
Comment (nik9000) [inline CONTRIBUTING.md:614]: Kind of like what we do with GC timings.

File: CONTRIBUTING.md (hunk 3 above)
Comment (gwbrown) [reply CONTRIBUTING.md]: Thanks, I'm good with the new wording for this document. I do think it might be a good idea to solidify this a bit more and re-evaluate the existing logs at some point, but that can be done as time and bandwidth allow.

File: CONTRIBUTING.md (hunk 5 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md:538]: > I guess to debug test failures?

Yeah either actual automated tests or cases where you can reproduce the problem in a non-production cluster but need the extra detail to see what's really happening. I use them particularly where whatever I'm chasing is super-rare, so I have to leave a test running for hours until it finally happens.

Maybe look at this the other way round: if (volume-wise) it's safe to log certain things in production, it can probably go to level `DEBUG`. If the volume is too high then it's going to be risky enabling them on a production cluster both in terms of the effects on the cluster itself and on whoever is consuming the logs. In any case we're not saying you can't use `TRACE` logs in prod, just that it might not be easy or even possible. Some customers may have policies that completely forbid use of the `TRACE` logging level in their production environments, for instance.

File: CONTRIBUTING.md (hunk 7 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md:605]: Sure, the most common place I use this is where the exception is almost always some boring network problem, like here:

https://github.com/elastic/elasticsearch/blob/68817d7ca29c264b3ea3f766737d81e2ebb4028c/server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java#L172-L176

Logging the stack trace here doesn't help, but it does generate a lot of noise when there's a network partition and that leads folks to report _just this message_ when asking for help.

On reflection maybe we should make this decision by looking at the exception too, only suppressing the stack trace if it really is a boring network problem.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -456,6 +456,188 @@ checks if contributed code does not have the appropriate license headers.
...
+running cluster. You should make sure that your contribution uses logging
+appropriately: log enough detail to inform users about key events and help them
+understand what happened when things go wrong, without logging so much detail
+that the logs fill up with noise and the useful signal is lost.
+
+Elasticsearch uses Log4J for logging. In most cases you should log via a
+`Logger` named after the class that is writing the log messages, which you can
+do by declaring a static field of the class as follows:
+
+    private static final Logger logger = LogManager.getLogger(MyClassName.class);
```
Comment (tvernum) [inline CONTRIBUTING.md:471]: In 31cd3b8a10fc @nik9000 introduced a lot of new loggers with

```suggestion
    private static final Logger logger = LogManager.getLogger();
```

which has the advantage of being concise & copy-paste safe, but the disadvantage that it has to inspect the callstack to do so.

Can we decide which version we prefer?

File: CONTRIBUTING.md (hunk 5 above)
Comment (tvernum) [reply CONTRIBUTING.md:538]: I think the phrase

> you would expect to be useful

is key here, and somewhat subjective.  
It can range from "If I try hard enough, I could come up with some exceptional scenario under which this information might help me with diagnosis" to "when this very common problem happens, I need to know this information in order to diagnose it".

For the latter, the ideal would be to detect the problematic scenario directly, and log at `WARN`.

But I wouldn't assume that "I can come up with a situation under which this might be useful" is the measure we're looking for. Because that's almost always going to be true.

I have no suggestions, and I don't even think there's anything wrong with the proposed text, I just think the way that different people read that phrase will affect whether they see it as too strict.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -456,6 +456,188 @@ checks if contributed code does not have the appropriate license headers.
...
+these messages to construct a high-level timeline of events leading up to an
+incident.
+
+This level is enabled by default so its target audience is the general
+population of users and administrators. You should use user-facing terminology
+and ensure that messages at this level are self-contained. In general you
+shouldn't log unusual events, particularly exceptions with stack traces, at
+`INFO` level. If the event is relatively benign then use `DEBUG`, whereas if
+the user should be notified then use `WARN`.
+
```
Comment (tvernum) [inline CONTRIBUTING.md:581]: I wonder whether it's worth commenting here that some readers are overly enthusiastic at interpreting info (or even debug/trace) messages as errors. Thus, it is important not to make the message sound like an error condition if it is not intended to be one.

```suggestion

It is often the case that people who read these logs are trying to determine why
their node is not behaving the way they expect - that is, they're in trouble
shooting mode. This can lead them to read more into a log message than the
developer intended. If your log message sounds like an error, then some
administrators will interpret it as one, even if it is logged at `INFO` level.
Where possible, `INFO` messages should prefer softer language such
as _"Did not find ..."_ rather than "Failed to find ..."_.

```

File: CONTRIBUTING.md (hunk 7 above)
Comment (tvernum) [reply CONTRIBUTING.md:605]: The other case is where the error is reported to the client, but we want to record it in the logs as well.

That's rare (we don't usually log client errors), but if it's an administrative action that's failed, sometimes we want the logs to mention it, but having a full stack trace each time isn't applicable, except for debug purposes.

File: CONTRIBUTING.md (hunk 5 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md]: ++ I did `s/expect to be useful/normally expect to be useful/` in 6942b041d2a13741434689ca599f49e811ceb858 and also added a note to discuss these things with reviewers when in doubt.

File: CONTRIBUTING.md (hunk 7 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md:616]: In 6942b041d2a13741434689ca599f49e811ceb858 I adjusted the guidance here to only suppress stack traces on boring exceptions. I opened https://github.com/elastic/elasticsearch/pull/75671 since the stack trace on `ReceiveTimeoutTransportException` is always boring anyway, and I'll align the code I quoted above with this new guidance at some point.

I also added more nuance in e1935754672401b360fed1328286b657dd3427b2 to the guidance about logging exceptions that directly result from client requests.

File: CONTRIBUTING.md (hunk 10 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md:591]: Thanks, that's a great point. I added (approximately) this in 6942b041d2a13741434689ca599f49e811ceb858

File: CONTRIBUTING.md (hunk 9 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md:471]: TIL, I didn't realise this did the right thing with subclasses, but it does. I don't have a strong opinion. What's the problem with inspecting the call stack?

File: CONTRIBUTING.md (hunk 6 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md:567]: We already mention thinking about (a) volume and (b) self-contained messages vs sequences. What sort of thing were you thinking of here?

File: CONTRIBUTING.md (hunk 9 above)
Comment (nik9000) [reply CONTRIBUTING.md:471]: Huh. I think I don't particularly like that construct any more. It feels kind of wasteful for it to do a bunch of stuff with the stack to get the name of the logger to save you a couple of characters in the reference. I tend not to do it now. But I've never benchmarked it. If it's fast then I'm all for it.

I'll check the performance of the two. I think I wouldn't mind a little performance hit but but I don't want us to slow down start up to build loggers. I'll check!

File: CONTRIBUTING.md (hunk 6 above)
Comment (nik9000) [reply CONTRIBUTING.md:567]: I was mostly thinking about concrete examples. Maybe with the whole structure of the document we could add some examples at the end and explain the reasoning. Sort of giving an example of applying the rules you've laid out. Its fine not to have it though.

File: CONTRIBUTING.md (hunk 7 above)
Comment (nik9000) [reply CONTRIBUTING.md:616]: Thanks!

File: CONTRIBUTING.md (hunk 9 above)
Comment (nik9000) [reply CONTRIBUTING.md:471]: I benchmarked them and the no-arg version is about 38 times slower. About 1.7-ish
microseconds:

```
Benchmark                       Mode  Cnt     Score    Error  Units
LogManagerBenchmark.inferClass  avgt   10  1763.708 ± 14.058  ns/op
LogManagerBenchmark.passClass   avgt   10    46.341 ±  0.539  ns/op
```

I'm sure there's issues with it, but this is what I used:

```
@Fork(2)
@Warmup(iterations = 10)
@Measurement(iterations = 5)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
public class LogManagerBenchmark {
    @Benchmark
    public Object passClass() {
        return LogManager.getLogger(LogManagerBenchmark.class);
    }

    @Benchmark
    public Object inferClass() {
        return LogManager.getLogger();
    }
}
```


So, if we do the no-arg version of `getLogger` in every java file we'll spend
an extra 27 milliseconds on logger initialization when we load the classes.
That seems fine. As long as the logger is shared.

File: CONTRIBUTING.md (hunk 6 above)
Comment (DaveCTurner) [reply CONTRIBUTING.md:605]: ++ I added some examples in 5cf06e4f1c3560edfee7c4f6e5d443fbb27b2d15.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -472,6 +472,236 @@ the [`Types.forciblyCast`](libs/core/src/main/java/org/elasticsearch/core/Types.
...
+On the other hand, users may be familiar with the `ERROR` level from elsewhere.
+Log4J for instance documents this level as meaning "an error in the
+application, possibly recoverable". The implication here is that the error is
+possibly _not_ recoverable too, and we do encounter users that get very worried
+by logs at `ERROR` level for this reason.
+
+Therefore you should try and avoid logging at `ERROR` level unless the error
+really does indicate that Elasticsearch is now running in a degraded state from
+which it will not recover. Errors like this should be very rare. When in doubt,
+prefer `WARN`.
```
Comment (jpountz) [inline CONTRIBUTING.md:703]: ERROR is the only one that doesn't have an example, I think it could be helpful since you described that it should be reserved to rare situations.

**Review by jpountz — APPROVED** [review]
Great write-up!

**Review by nik9000 — COMMENTED** [review]
:+1: Thanks!
